app/services/notifications/journal_wp_notification_service.rb

```diff
@@ -45,105 +45,88 @@ def call(journal, send_notifications)
+      }.merge(channel_attributes(channels))
 
       Notifications::CreateService
         .new(user: user)
         .call(notification_attributes)
     end
 
-    def channel_attributes(recipient, reason)
-      key =
-        case reason
-        when :subscribed
-          :all
-        else
-          reason
-        end
-
-      recipient.notification_settings.map do |setting|
-        channel =
-          case setting.channel
-          when 'mail'
-            :read_email
-          when 'in_app'
-            :read_ian
-          else
-            raise "Unknown notification channel"
-          end
+    def channel_attributes(channels)
+      {
+        read_email: channels.include?('mail') ? false : nil,
+        read_ian: channels.include?('in_app') ? false : nil
 
-        [channel, setting[key] ? false : nil]
-      end.to_h
+      }
     end
 
     def notification_receivers(work_package, journal)
-      seen = mentioned(journal).each do |user|
-        yield(user, :mentioned)
-      end
-
-      seen += involved(journal, seen).each do |user|
-        yield(user, :involved)
-      end
+      receivers = receivers_hash
 
-      seen += subscribed(journal, seen).each do |user|
-        yield(user, :subscribed)
-      end
+      add_receiver(receivers, mentioned(journal), :mentioned)
+      add_receiver(receivers, involved(journal), :involved)
+      add_receiver(receivers, watched(work_package), :watched)
+      add_receiver(receivers, subscribed(journal), :subscribed)
 
-      watched(work_package, seen).each do |user|
-        yield(user, :watched)
-      end
+      receivers
     end
 
     def mentioned(journal)
       allowed_and_unique(mentioned_ids(journal),
                          journal.data.project,
-                         [])
+                         :mentioned)
     end
 
-    def involved(journal, seen)
+    def involved(journal)
       scope = User
         .where(id: group_or_user_ids(journal.data.assigned_to))
         .or(User.where(id: group_or_user_ids(journal.data.responsible)))
 
       allowed_and_unique(scope,
                          journal.data.project,
-                         seen)
+                         :involved)
     end
 
-    def subscribed(journal, seen)
+    def subscribed(journal)
       project = journal.data.project
 
       allowed_and_unique(User.notified_on_all(project),
                          project,
-                         seen)
+                         :all)
     end
 
-    def watched(work_package, seen)
+    def watched(work_package)
       allowed_and_unique(work_package.watcher_users,
                          work_package.project,
-                         seen)
+                         :watched)
     end
 
-    def allowed_and_unique(scope, project, seen = [])
-      scope
-        .where(id: User.allowed(:view_work_packages, project))
-        .where.not(id: seen.map(&:id))
-        .not_builtin
+    def allowed_and_unique(user_scope, project, reason)
```

Comment:
This method name lacks some hint for the return value. 
```suggestion
    def notification_settings_if_allowed(user_scope, project, reason)
```

Reply:
Renamed in #9430